Raylib Driverstation Current Progress

PhocaeaUnitas opened the pull request:
See issue #8
See issue #17

PhocaeaUnitas committed:
Added current driverstation progress

See: https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/issues/8
See: https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/issues/17

yameatmeyourdead commented on src/driverstation/.gitignore:
Why in this commit :)

yameatmeyourdead commented on src/driverstation/build.sh:
git ignore this

yameatmeyourdead commented on src/driverstation/run.sh:
git ignore this

PhocaeaUnitas committed:
Removed build scripts;
Added `.clangd` to .gitignore;

PhocaeaUnitas committed:
Removed build scripts;
Added `.clangd` to .gitignore;

yameatmeyourdead commented on src/driverstation/include/ssh_ui.hpp:
```diff
@@ -0,0 +1,771 @@
+				default:
+					// TODO: Temporary
+					std::cout << "Unknown escape code: \\x1B[";
+					for(int i = 0; i < this->current_csi_parameters.size(); i++){
```
cast i to size_t to prevent compiler warning about signed/unsigned comparison

PhocaeaUnitas replied:
See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e4d02ec2f011f01f0159acd6f4bd5bf98eafde3389acf8a6ceb66a36aeac33a0L710-R710)

yameatmeyourdead reviewed (commented):
Pretty good so far. Most of my current requests for changes deal with compiler warnings. Try developing with `-Wall -Wextra -Wpendantic` to catch these. Some of these do actually have performance impacts when compiling with optimization level `-O3`.

PhocaeaUnitas committed:
Added model assets (were ignored by outer `.gitignore` as compilation artifacts);
Replaced hex character literals with hex numeric literals;
Corrected variable types on vector-iterating for loops;
Removed unused variables;
Removed now-ignored `.clangd` from index;
Flagged unused parameters as `[[maybe_unused]]`;
Flagged intentional fallthroughs with `[[fallthrough]]`;

PhocaeaUnitas commented on src/driverstation/include/ssh_ui.hpp:
```diff
@@ -0,0 +1,771 @@
+											uint8_t n = parameters[i + 2].value;
+
+											// NOTE: The standard palette options modify the `n` parameter such that the next pass actually handles the color application
+											if(n >= 0 && n <= 7){  // Standard foreground color
```
See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e4d02ec2f011f01f0159acd6f4bd5bf98eafde3389acf8a6ceb66a36aeac33a0L604-R604)

PhocaeaUnitas commented on src/driverstation/include/ssh_ui.hpp:
```diff
@@ -0,0 +1,771 @@
+							break;
+						case TextMode::AWAITING_STRING_TERMINATOR:
+							switch(current_character){
+								case '0x1B':  // Escape, possibly String Terminator
```
See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e4d02ec2f011f01f0159acd6f4bd5bf98eafde3389acf8a6ceb66a36aeac33a0L257-R257)

PhocaeaUnitas commented on src/driverstation/include/ssh_ui.hpp:
```diff
@@ -0,0 +1,771 @@
+								case '\a':  // For historical reasons, xterm can end OSCs with a bell
+									this->current_text_mode = TextMode::DEFAULT;
+									break;
+								case '0x1B':  // Escape, possibly String Terminator
```
See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e4d02ec2f011f01f0159acd6f4bd5bf98eafde3389acf8a6ceb66a36aeac33a0L240-R240)

PhocaeaUnitas commented on src/driverstation/include/dynamic_texture.hpp:
```diff
@@ -0,0 +1,395 @@
+			return this->render_texture.texture;
+		}
+
+		void on_click(float x, float y) override{}
```
See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-0e6222c0c79b21dc1af73b516d394ec72339ca3798e216ea618bd23100b694ebL62-R62)

PhocaeaUnitas commented on src/driverstation/include/dynamic_texture.hpp:
```diff
@@ -0,0 +1,395 @@
+			return this->texture;
+		}
+
+		void on_click(float x, float y) override{}
```
See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-0e6222c0c79b21dc1af73b516d394ec72339ca3798e216ea618bd23100b694ebL200-R200)

PhocaeaUnitas commented on src/driverstation/include/estop_ui.hpp:
```diff
@@ -0,0 +1,57 @@
+			return dynamic_texture::DynamicRenderTexture2D::current_texture();
+		}
+
+		void on_click(float x, float y) override{
```
See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e8a70cdef6d2873c987c607e6e340a3f6437fe355da6f145cd332d9797048e3eL32-R32)

PhocaeaUnitas commented on src/driverstation/include/ssh_interface.hpp:
```diff
@@ -0,0 +1,235 @@
+		                    "automatically created.\n");
+
+		            /* FALL THROUGH to SSH_SERVER_NOT_KNOWN behavior */
+
```
See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-21afed6ef9ec3a758c8bec0a5380ca1cf5cd5937d330b53d0cefa4efa694bdf7L71-R71)

PhocaeaUnitas commented on src/driverstation/include/ssh_ui.hpp:
```diff
@@ -0,0 +1,771 @@
+						if(parameters.size() > 0) characters_to_delete = nonZeroOr(parameters[0].value, 1);
+
+						float width = this->history_rectangle.width - (characters_to_delete * this->character_spacing.x);
+						Texture2D texture = this->current_texture();
```
See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e4d02ec2f011f01f0159acd6f4bd5bf98eafde3389acf8a6ceb66a36aeac33a0L553-L554)

PhocaeaUnitas commented on src/driverstation/include/ssh_ui.hpp:
```diff
@@ -0,0 +1,771 @@
+					break;
+				case 'm':  // Select Graphic Rendition parameters
+					// TODO: Rework to make use of CsiParameter.type property
+					for(int i = 0; i < parameters.size(); i += 2){  // NOTE: += 2 to skip semicolons separating parameters
```
See [40be54a](https://github.com/Colorado-School-of-Mines-Robotics-Club/MATE_ROV_Programming/commit/40be54aa21dde2c1cf1b5c6848df74e4cb01a223#diff-e4d02ec2f011f01f0159acd6f4bd5bf98eafde3389acf8a6ceb66a36aeac33a0L581-R581)

PhocaeaUnitas commented on src/driverstation/include/driverstation_ui.hpp:
```diff
@@ -0,0 +1,156 @@
+		protected:
+			std::unique_ptr<dynamic_texture::MultiCellDynamicTextureGrid2D<12, 12>> multi_cell;
+			std::shared_ptr<gui::SshDisplay> ssh_display;
+			SshConnectPopupMode ssh_connect_popup_mode;
```
Use C++20 and I'll use modules instead

PhocaeaUnitas committed:
`DriverstationUi::setCameraImage` can be used to set image data of individual cameras;
`DriverstationUi::registerEstopCallback` can be used to add a callback function to be called when the estop is triggered;

yameatmeyourdead commented:
Closing this, this branch has been replaced by driverstation-rosified. Will not merge.
